src/aics-image-viewer/components/ViewerStateProvider/ResetStateProvider.ts

```diff
@@ -0,0 +1,151 @@
+import { isEqual } from "lodash";
+
+import { ChannelState, ViewerState, ViewerStateContextType } from "./types";
+import {
+  getDefaultViewerState,
+  getDefaultCameraState,
+  getDefaultChannelState,
+  getDefaultViewerChannelSettings,
+  getDefaultChannelColor,
+} from "../../shared/constants";
+import { ViewMode } from "../../shared/enums";
+import {
+  overrideViewerState,
+  overrideChannelStates,
+  getEnabledChannelIndices,
+  initializeOneChannelSetting,
+} from "../../shared/utils/viewerState";
+import { ViewerChannelSettings } from "../../shared/utils/viewerChannelSettings";
+
+/**
+ * Provides reset functionality for the viewer and channel states.
+ */
+export default class ResetStateProvider {
+  savedViewerState: Partial<ViewerState>;
+  savedViewerChannelSettings: ViewerChannelSettings | undefined;
+  useDefaultViewerChannelSettings: boolean;
+
+  channelsToReset: Set<number>;
+  channelsToResetOnLoad: Set<number>;
+
+  ref: React.MutableRefObject<ViewerStateContextType>;
+
+  constructor(viewerStateRef: React.MutableRefObject<ViewerStateContextType>) {
+    this.ref = viewerStateRef;
+
+    this.savedViewerState = {};
+    this.channelsToReset = new Set();
+    this.channelsToResetOnLoad = new Set();
+    this.savedViewerChannelSettings = undefined;
+    this.useDefaultViewerChannelSettings = false;
+
+    this.resetToSavedViewerState = this.resetToSavedViewerState.bind(this);
+    this.resetToDefaultViewerState = this.resetToDefaultViewerState.bind(this);
+    this.resetToState = this.resetToState.bind(this);
+    this.setSavedViewerState = this.setSavedViewerState.bind(this);
+    this.setSavedViewerChannelSettings = this.setSavedViewerChannelSettings.bind(this);
+    this.getChannelsAwaitingReset = this.getChannelsAwaitingReset.bind(this);
+    this.getChannelsAwaitingResetOnLoad = this.getChannelsAwaitingResetOnLoad.bind(this);
+    this.onResetChannel = this.onResetChannel.bind(this);
+    this.getCurrentViewerChannelSettings = this.getCurrentViewerChannelSettings.bind(this);
+  }
+
+  // Setup Callbacks ////////////////////////////////////////////////////////////////////
+
+  public setSavedViewerState(state: Partial<ViewerState>): void {
+    this.savedViewerState = state;
+  }
+
+  public setSavedViewerChannelSettings(settings: ViewerChannelSettings | undefined): void {
```

Comment:
Undefined is passed in if the viewer is used in embedded mode but there are no initial settings/props passed in. This can also happen when opening a URL with no parameters set, I believe.